### PR TITLE
postgresql: always create the bin directory in postgresqlAndPlugins

### LINKED ISSUE #22653: Postgresql extraPlugins require /bin dir to exist in extension directory
## Issue description

I was packaging pgjwt postgresql extension and found small issue. When packaged extension don't have a `/bin` directory, buildEnv function which builds new environment for postgresql-with-plugins fails due to `rm` permission issues.

It looks like `buildEnv` calculate output based on /bin directories, when postgresql extension does not have /bin, `buildEnv` uses "old" postgresql-with-plugins derivation, but in the built script, there is rm command to remove "/bin/*" and copy files again. `rm` fails because /nix/store doesn't allow it to write/delete anything in already built derivation.

Attached pull request with "patched" extension:
https://github.com/NixOS/nixpkgs/pull/22644

### Steps to reproduce

Add this to nixos configuration and rebuild:
```
service.postgresql.enabled=true;
service.postgresql.extraPlugins = [X];
# where X is any derivation without a /bin directory
```

## Technical details

* System: (NixOS: 17.03pre99626.f1ba2c8)
* Nix version: nix-env (Nix) 1.11.6
* Nixpkgs version: 17.03pre100479.f7b7d8

### PR DESCRIPTION
**Pull request by marsam**
This is needed because some PostgreSQL plugins don't have a bin
directory. If only these plugins are listed in cfg.extraPlugins
buildEnv will turn $out/bin into a symbolic link to ${pg}/bin. Lateron
we try to rm $out/bin/{pg_config,postgres,pg_ctl} which will then fail
because $out/bin will be read-only.

Rebased from https://github.com/NixOS/nixpkgs/pull/44379, which didn't get merged
Closes https://github.com/NixOS/nixpkgs/issues/22653

backport please

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by aanderse** [comment]
@marsam Is this PR still valid? Are you just waiting for review from someone like @thoughtpolice, @danbst, or @aszlig?

**Comment by marsam** [comment]
>Is this PR still valid?

yes, is still valid. I'd really appreciate a review of the PostgreSQL maintainers

**Review by 7c6f434c — APPROVED** [review]

**Review by danbst — APPROVED** [review]

**Comment by aanderse** [comment]
Thanks team!

**Pull request merged**
